THP documentation (#6196)
trezor/trezor-firmware: docs/common/thp/specification.md

>
> — Bluetooth Core_v5.3 specification, section 5.4.5, page 275
>

### BLE packet size

Trezor utilizes packets of 244 bytes in size.

# Transport layer

The transport layer performs the segmentation of transport payloads into multiple packets that are transmitted over the data transfer layer. It is possible for several applications to communicate with Trezor using distinct transport connections, which are referred to as *channels*. The transport layer is responsible for the establishment and release of channels, multiplexing channels onto the data transfer layer, and channel synchronization.

## Channel identifier

Channels are identified by a 16-bit *channel identifier* (CID). The CID `0xFFFF` is reserved for the broadcast channel, which is used for channel allocation requests and responses. CIDs from `0xFFF0` to `0xFFFE`, and `0x0000`, are reserved for future use. Trezor allocates the remaining CIDs to individual host applications.

## Transport packet structure

Transport packets start with a *control byte* to determine how to handle the received packet. The recognized control byte values (with appropriate masks) are denoted in the following table.

> 

Reviewer: I am confused by the Xs and the masks. I can somewhat guess what it is used for but I think a sentece desribing it would be nice

| ack                           | `0010X000` | 0xF7 | 0x20 | Acknowledgement message.                           | L 2.5 - Synchronization                   |
| handshake_init_request        | `000XX000` | 0xE7 | 0x00 | First message in the handshake process             | L 3 - [Handshake phase](#handshake-phase) |
| handshake_init_response       | `000XX001` | 0xE7 | 0x01 | Second message in the handshake process            | L 3 - [Handshake phase](#handshake-phase) |
| handshake_completion_request  | `000XX010` | 0xE7 | 0x02 | Third message in the handshake process             | L 3 - [Handshake phase](#handshake-phase) |
| handshake_completion_response | `000XX011` | 0xE7 | 0x03 | Fourth message in the handshake process            | L 3 - [Handshake phase](#handshake-phase) |
| encrypted_transport           | `000XX100` | 0xE7 | 0x04 | Noise message, i.e. encrypted application message. | L 3                                       |

The transport layer distinguishes two basic types of transport packets: *initiation packets* and *continuation packets*. The highest bit of the control byte specifies whether the packet is an *initiation packet* or a *continuation packet.*

> - *type* = 0  *→ initiation packet*

Author: Technically yes. The point is that most packet (message) properties are specified only in the first (initiation) packet. Example: when a message is an `encrypted_transport` message, you specify it only in the first packet. The following packets are "continuation" packets - they do not mention the `encrypted_transport` again.

| **Offset** | **Size** | **Field name**   | **Description**                                                                                                |
|:----------:|:--------:|:-----------------|:---------------------------------------------------------------------------------------------------------------|
| 0          | 1        | *control_byte*   | The control byte set to 0x80.                                                                                  |
| 1          | 2        | *cid*            | Channel identifier as a 16-bit integer in big-endian byte order.                                               |
| 3          | *n*−3    | *packet_payload* | Continuation part of the *transport payload with CRC*. Padded with null bytes in the last continuation packet. |

The seven least significant bits of the control byte in a continuation packet are reserved for future use. The reserved bits should be set to zero when transmitting a continuation packet, and ignored when receiving a continuation packet.

## Transport layer structure

The transport layer is further divided into five sub-layers.

> ![THP layers - detail](./images/layersTHP-v42.svg)

Reviewer: The sizes in this image are confusing. On the left side there is 1 bit = 2 bytes (line L2.2) while 1 byte (line L2.3) > 2 bytes.

### Channel multiplexing layer

> Multiplexing enables one data transfer connection to support more than one channel. This layer manages the identification of the channel for each packet transferred over the data transfer connection, in order to ensure that packet payloads from the various multiplexed channels are not mixed. The receiver groups the packets based on their *channel identifier* (CID) field. The channel multiplexing layer does not provide any packet filtering or error handling. (TODO add relevant part of session locking)

Author: Yes, it looks at the 2nd and 3rd byte (CID is 2 bytes in length). 

The reason why the CID is not at the beginning of a packet is the ease of implementation/description. THP must be able to identify a CodecV1 packet - which can be done by examining the first byte of the packet. When the control byte is the first byte, this can be done easily and fairly naturally. However, if CID was the first part of the packet, it would be necessary to:
 - prohibit the use of CIDs that have the first byte same as is the CodecV1 value.
 - compare the first byte of the CID with the CodecV1 value

So even though we look at the CID first, it makes sense to not have a packet start with it.

If an initiation packet is received on a channel CID2 while reassembling the *transport payload with CRC* on a different channel CID1, then the receiver should reassemble both transport payloads with CRC in parallel independently for each channel.

If the receiver is not able to reassemble multiple payloads in parallel, then it should measure the time that elapsed since the last packet was received on channel CID1. If that time exceeds MIN_CONTINUATION_WAIT_MS, it should abort the reassembly of the *transport payload with CRC*, discard any previously reassembled data and begin reassembling the transport payload on channel CID2. If MIN_CONTINUATION_WAIT_MS is not exceeded, then it should notify the allocation layer to send a TRANSPORT_BUSY error on channel CID2.

If a continuation packet is received on a given channel when none is expected on that channel, then the receiver will discard the continuation packet.

### Error detection layer

> A cyclic redundancy check (CRC) error detection code is used to detect data corruption or the loss of continuation packets by the data transfer layer. The algorithm used to compute the CRC is CRC-32-IEEE, which utilizes the polynomial 0x04C11DB7 with its reversed form equal to 0xEDB88320. The CRC is computed over the header of the initiation packet and the transport payload as outlined in the table below. The result is encoded as 4 bytes in big-endian byte order and appended to the transport payload, which is referred to as *transport payload with CRC*.

Reviewer: nit: would it make sense to add a reference to some specification of CRC or CRC-32-IEEE? Either something like (CRC)[https://en.wikipedia.org/wiki/Cyclic_redundancy_check] or a reference to the  CRC-32-IEEEalgorithm.

Reviewer: Also I am not sure if sentence "which utilizes the polynomial 0x04C11DB7 with its reversed form equal to 0xEDB88320" isn't unnecessarily confusing here. Are the values not standard? Or are they referenced later in the specification?

| 3          | 2            | *length*            | Total size of the *transport payload with CRC* in bytes. Encoded as a 16-bit integer in big-endian byte order. |
| 5          | *length* – 4 | *transport_payload* | The transport payload. (Does not include CRC or padding.)                                                      |

The receiver validates the CRC. If invalid, then the transport payload is discarded.

### Allocation layer

The allocation layer handles the establishment and release of channels and transport error handling. Host applications initiate communication with Trezor by requesting the establishment of a channel. The Trezor is responsible for the allocation of a unique channel identifier to each host application.

If Trezor receives a transport payload with a channel identifier that is not allocated, it will respond with an UNALLOCATED_CHANNEL error using the same channel identifier. Note: Implementation can respond to not allocated channel already after receiving an initiation packet (with unknown/unallocated CID).

> A host application will ignore any data transmitted on a channel that has not been allocated to it.

Reviewer: Do we enforce this somehow or do we hope the host does this on its own?

    required uint32 protocol_version_major = 3;       // The major version of the communication protocol used by the firmware.
    required uint32 protocol_version_minor = 4;       // The minor version of the communication protocol used by the firmware.
    repeated ThpPairingMethod pairing_methods = 5;    // The pairing methods supported by the Trezor.
}
```

The pairing methods are further described in the [Pairing phase](#pairing-phase).

```protobuf
// Numeric identifiers of pairing methods.
enum ThpPairingMethod {
    SkipPairing = 1;          // Trust without MITM protection.
    CodeEntry = 2;            // User types code displayed on Trezor into the host application.
    QrCode = 3;               // User scans code displayed on Trezor into host application.
    NFC = 4;                  // Trezor and host application exchange authentication secrets via NFC.
}
```

The supported pairing methods are not implied by the model because some may be disabled by user settings or device mode. Trezor may also indicate a different set of pairing methods depending on the underlying means of data transfer, e.g., USB and Bluetooth.

> `TransportError` uses the `transport_error` control byte and has the following transport payload:

Reviewer: It seems it is the means to communicate various errors, right? If so, could we add a line explaining that?

Recognized transport error codes are:

| **Code** | **Name**            | **Description**                                                                                                                                         |
|:--------:|:--------------------|:--------------------------------------------------------------------------------------------------------------------------------------------------------|
| 0        |                     | Reserved for future use.                                                                                                                                |
| 1        | TRANSPORT_BUSY      | Issued by a recipient when the transport layer is busy reassembling a message on another channel.                                                       |
| 2        | UNALLOCATED_CHANNEL | Issued by Trezor in response to messages that have a channel identifier that is not allocated.                                                          |
| 3        | DECRYPTION_FAILED   | Issued by Trezor in response to messages that have an invalid authentication tag. Decryption error results in termination of the channel.               |

> | 4        |                     | Reserved for future use.                                                                                                                                |

Author: It is completely your business 😄 

The code 4 was briefly used in development, but it was later discarded. The DEVICE_LOCKED 5 error was already used when we realized the error with code 4 is not - therefore it was not worth shuffling the error codes.

The code 4 can be safely used in the future, in case a new error is needed.

> | 5        | DEVICE_LOCKED       | Issued by Trezor in response to handshake messages (`HandshakeInitiationRequest`, `HandshakeContinuationRequest`) that are sent to a **locked** device. |

Author: Reference added here: https://github.com/trezor/trezor-firmware/pull/6196/changes/847f7c9d63f234d925f6cf3a8674424e93b31c81

When the sender receives a TRANSPORT_BUSY error, it should notify the segmenting layer to cease transmitting packets and notify the synchronization layer to increase the timeout for the next retransmission of the transport payload by a random value in the range [0, MAX_BUSY_BACKOFF_MS].

When the host receives an UNALLOCATED_CHANNEL or DECRYPTION_FAILED error, it should consider the channel released, destroy any channel context, and establish a new channel with Trezor.

In addition, THP supports sending and receiving keep-alive messages (`Ping` and `Pong`) over a channel. `Ping` is sent with a random 8-byte nonce, to make sure the peer is available and responsive when `Pong` response is received (with the same nonce). Since it is a single packet message, no payload reassembly is required. This mechanism can be used by both sides (for example, the device can check that the channel is still responsive, and close it if needed).

### Synchronization layer

> The synchronization layer handles acknowledgement messages, which allow the receiver to inform the sender of the receipt of each transport payload. If acknowledgement is not received within a defined time, the sender infers the non-receipt of a transport payload and the need to re-transmit it. The sender may re-transmit the transport payload indefinitely. However, implementations can set a limit on the number of retransmissions (MAX_RETRANSMISSION_COUNT) to prevent power drainage or blocking of system resources. In particular, if the sender needs to transmit a transport payload on another channel and does not have the system resources to do so in parallel on both channels, then it should terminate retransmission on the unresponsive channel as long as at least MIN_RETRANSMISSION_COUNT retransmissions have been attempted. When retransmission is

Author: No, it is not. The time is not protocol-defined - it is implementation-defined. It was already changed several times - it is currently dynamic and it also depends on the responsiveness of the channel / interface used.

I am unsure whether it should be specified within the protocol docs.

- **terminated** without having received acknowledgement, the sender should release the unresponsive channel.
- **interrupted** (e.g. by losing BLE connection), the sender could continue the retransmission when the connection is (possibly) restored (or terminate the channel).

Sender should not assume a message to be received without a receipt of an acknowledgement message.

The primary function of the synchronization layer is to work as the Alternating Bit Protocol (ABP) to ensure reliable communication. As such, it handles sequence numbers, ACK messages, retransmissions, and timeouts. Secondly, it provides the higher layers (e.g., the Secure channel layer) with data payloads (and notifications).

It is possible that some packets are lost by the Data transfer layer (L1). In order to guarantee reliable communication, THP uses the *Alternating Bit Protocol* in combination with a *CRC32* checksum.

### Alternating Bit Protocol

> The *Alternating Bit Protocol* (ABP) is a protocol between a sender and a receiver over an unreliable channel. The channel is unreliable in the sense that it can discard or duplicate messages, but it cannot change their order or content. The protocol guarantees the eventual and non-duplicative delivery of messages.

Reviewer: I would consider adding some reference to emphasize that this is a standard protocol and not a custom solution

> After sending a message, the sender doesn't send any further messages until it receives an *acknowledgement message* (ACK). After receiving a valid message, the receiver sends an ACK. If the ACK does not reach the sender before a timeout, the sender re-transmits the message. THP uses a 1-bit sequence number that is part of the header of the initialization packet. This sequence number alternates (from 0 to 1) in subsequent messages. When the receiver sends an ACK, it includes the sequence number of the message it received.

Reviewer: Just to check if I understand it, if two subsequent messages get lost, the protocol does not recognize the loss?

The receiver can detect duplicated frames by checking if the message sequence numbers alternate. If two subsequent messages have the same sequence number, they are duplicates, and the second message is discarded. Similarly, if two subsequent ACKs reference the same sequence number, they are acknowledging the same message.

The situation is described by the following state machine.

> ### ABP State machine

Author: It was intentional, but I will change the structure. The document was originally written in Notion, which supports only 3 heading levels. Markdown (including GitHub markdown) supports up to 6 heading levels - docs.github.com/en/get-started/writing-on-github/getting-started-with-writing-and-formatting-on-github/basic-writing-and-formatting-syntax#headings

When the structure is changed, I will add the commit link into this thread.

> ### ABP State machine

Author: heading level adjusted here:
https://github.com/trezor/trezor-firmware/commit/be31c1e57ba8cd08ec05012918829b850292fe70

    - Transition to the state S0a.
- When the acknowledgment message is received and its sequence number equals 1, take the following actions:
    - Consider the previously sent message to be delivered.
    - Transition to the state S0b.

The behavior of the sender in the state S0b is defined as follows:

- If there is a message that needs to be sent, take the following actions:
    - Send the message with the sequence number equal to 1.
    - Switch to state S1a.
    - Reset and start the stopwatch.
- When the acknowledgment message is received and its sequence number equals 1, take the following actions:
    - Consider the message to be a duplicate acknowledgment of the previously sent message.
    - Transition to the state S0b.
- When the acknowledgment message is received and its sequence number equals 0, take the following actions:
    - Consider the message to be an acknowledgment received out of order.
    - Transition to the state S0b.

The behavior of the sender in the state S1 is analogous. The full finite-state transducer of the sender is depicted in the following diagram:

```mermaid
flowchart LR
    S0a["S0a"]
    S0b["S0b"]
    S1a["S1a"]
    S1b["S1b"]

>     S0a -- receive Ack(Seq=0) --> S0a

Author: Improved

    S0b -- send Message(Seq=0, Payload=payload0) --> S1a

>     S0a -- after timeout\nresend Message(Seq=1, Payload=payload1) --> S0a:::dashed

Reviewer: also the newlines do not seem to be rendering correctly

    S1a -- after timeout\nresend Message(Seq=0, Payload=payload0) --> S1a:::dashed
    S1b -- receive Ack(Seq=*) --> S1b

    classDef dashed stroke-dasharray: 5 5;
```

The receiver alternates between states R0 and R1. The receiver starts in the state R0.

The behavior of the receiver in the state R0 is defined as follows:

- When the message is received and its sequence number equals 0, take the following actions:
    - Submits the received message to further processing.
    - Reply with an acknowledgement message with the sequence number equal to 0.
    - Transition to the state R1.
- When the message is received and its sequence number equals 1, take the following actions:
    - Consider the message to be a duplicate of the previously received message.
    - Reply with an acknowledgement message with the sequence number equal to 1.
    - Transition to the state R1.

The behavior of the receiver in the state R1 is analogous. The full finite-state transducer of the sender is depicted in the following diagram:

```mermaid
flowchart LR
    R1["R1"]
    R2["R2"]

    R1 -- in_synchronization: Message(Seq=1, Payload=payload); out_synchronization: Ack(Seq=1) --> R1

>     R1 -- in_synchronization: Message(Seq=0, Payload=payload); out_synchronization: Ack(Seq=0); out: Message(Payload=payload) --> R2

Reviewer: I see these two boxes overlay each other

    optional bytes host_pairing_credential = 1;       // Host's pairing credential
}
```

### Handshake completion response (HandshakeCompletionResponse)

The message is sent from the Trezor to the host. It is sent as an encrypted payload.

| Offset | Size | Field name               | Description                                |
|--------|------|--------------------------|--------------------------------------------|
| 0      | 17   | *encrypted_trezor_state* | encrypted and authenticated Trezor’s state |

## Trezor’s state machine

Let *device_properties* be the properties advertised by the Trezor in ChannelAllocationResponse.

Let (*trezor_static_privkey*, *trezor_static_pubkey*) be a X25519 key pair that persists Trezor restarts.

Let *cred_auth_key* be a symmetric 16-byte key that persists Trezor's restarts.

```mermaid
flowchart TD
  TH1 -- in:&nbsp;HandshakeInitiationRequest<br>out:&nbsp;HandshakeInitiationResponse --> TH2
  TH2 -- in:&nbsp;HandshakeCompletionRequest&nbsp;[paired]<br>out:&nbsp;HandshakeCompletionResponse --> TS["transport state"]
  TH2 -- in:&nbsp;HandshakeCompletionRequest&nbsp;[unpaired]<br>out:&nbsp;HandshakeCompletionResponse --> TP1
```

> ### State TH1

Reviewer: I understand it is not necessary but I would advise on adding some comments describing the states on high level. Like what are the main values that are computed and what this state "means"

Author: Tl;dr: I suggest keeping it as is, at least for now.

I would deffer adding comments to the states to future improvements. In general, the states are just a part of a state-machine(s). The state-machines are thought through and have meaning (handshake / pairing / ...) but the states emerged as a description of the machines. To give an example of what I mean: - the difference between  states TH1 and TH2 - is that the steps in state TH2 take part after steps TH1 and there was some input from the host in between. It is possible to retrospectively add some meaning what TH1 and TH2 mean. However, in general, the description is their content. There is a process "handshake" that has many steps taken by the host and the Trezor - and the states just divide them into smaller "units". I could, theoretically, write all of the steps in one place (HH0, TH1, HH1, TH2, HH2/HH3). It is better to have it separated, because steps taken by the Trezor are (mostly) irrelevant for the implementation of the host application.

The behavior of the Trezor in the state TH1(*device_properties*, *trezor_static_pubkey*, *trezor_static_privkey*) is defined as follows:

When the message HandshakeInitiationRequest(*host_ephemeral_pubkey*) is received, take the following actions:
1. Generate a new ephemeral X25519 key pair (*trezor_ephemeral_privkey*, *trezor_ephemeral_pubkey*).
2. Set *h* = SHA-256(*protocol_name* || *device_properties*).
3. Set *h* = SHA-256(*h* || *host_ephemeral_pubkey*).
4. Set *h* = SHA-256(*h* || *try_to_unlock*).
5. Set *h* = SHA-256(*h* || *trezor_ephemeral_pubkey*).
6. Set *ck*, *k* = HKDF(*protocol_name*, X25519(*trezor_ephemeral_privkey*, *host_ephemeral_pubkey*)).
7. Set *mask* = SHA-256(*trezor_static_pubkey* || *trezor_ephemeral_pubkey*).
8. Set *trezor_masked_static_pubkey* = X25519(*mask*, *trezor_static_pubkey*).

> 9. Set *encrypted_trezor_static_pubkey* = AES-GCM-ENCRYPT(*key*=*k*, *IV*=0^96 (96 bits), *ad*=*h*, *plaintext*=*trezor_masked_static_pubkey*).

Author: removed here: https://github.com/trezor/trezor-firmware/commit/6b1e876ac04bf844151a98ef44dbecec57f6697f

2. Set *h* = SHA-256(*h* || *encrypted_host_static_pubkey*).
3. Set *ck*, *k* = HKDF(*ck*, X25519(*trezor_ephemeral_privkey*, *host_static_pubkey*)).
4. Set *payload_binary* = AES-GCM-DECRYPT(*key*=*k*, *IV*=0^96, *ad*=*h*, *ciphertext*=*encrypted_payload*).
5. Set *h* = SHA-256(*h* || encrypted_payload).
6. Set *payload* = PROTOBUF-DECODE(*type*=HandshakeCompletionReqNoisePayload, *data*=*payload_binary*).
7. If *ValidateCredential*(*cred_auth_key*, *payload.host_pairing_credential*, *host_static_public_key*) == True, then set *trezor_state* = STATE_PAIRED (or STATE_PAIRED_AUTOCONNECT - see [Channel replacement](#channel-replacement)). Otherwise, set *trezor_state* = STATE_UNPAIRED.
8. Set *key_request*, *key_response* = HKDF(*ck*, *empty_string*).
9. Set *encrypted_trezor_state* = AES-GCM-ENCRYPT(*key*=*key_response*, *IV*=0^96, *ad*=*empty_string*, *plaintext*=*trezor_state*).
10. Send the message HandshakeCompletionResponse(*encrypted_trezor_state*) to the host.
11. Set *nonce_request* = 0 and *nonce_response* = 1.
12. Set *encryption_state* = (*key_request*, *key_response*, *nonce_request*, *nonce_response*).

> 13. If *trezor_state* == STATE_PAIRED, transition to the state TC1. If *trezor_state* == STATE_UNPAIRED, transition to the state TP1(*host_static_public_key*, *h*).

Reviewer: In the flowchart above, the "paired" branch goes to a "transport state". Is TC1 the same?

## Host’s state machine

Let (*host_static_privkey*, *host_static_pubkey*) be a X2551thpPairing9 key pair that persists host restarts.

Let *credentials* be a list of pairs consisting of Trezor’s static public key and the corresponding credential.

> Let *device_properties* be the Trezor’s properties advertised by the Trezor in ChannelAllocationResponse.

Reviewer: Just to be sure, ChannelAllocationResponse is not formatted by design? I remember the message names used to be formatted as code (like `ChannelAllocationResponse`)

    ```mermaid
    flowchart TD
      HH0 -- out:&nbsp;HandshakeInitiationRequest --> HH1
      HH1 -- [Trezor&nbsp;is&nbsp;known]<br>in:&nbsp;HandshakeInitiationResponse<br>out:&nbsp;HandshakeCompletionRequest&nbsp;[real&nbsp;key] --> HH2
      HH1 -- [Trezor&nbsp;is&nbsp;unknown]<br>in:&nbsp;HandshakeInitiationResponse<br>out:&nbsp;HandshakeCompletionRequest&nbsp;[dummy&nbsp;key] --> HH3
      HH3 -- in:&nbsp;HandshakeCompletionResponse [trezor&nbsp;is&nbsp;not&nbsp;paired] ---> HP0
      HH2 -- in:&nbsp;HandshakeCompletionResponse [trezor&nbsp;is&nbsp;paired] ---> HC1
      HH2 -- in:&nbsp;HandshakeCompletionResponse [trezor&nbsp;is&nbsp;not&nbsp;paired] ---> HP0
    ```

### State HH0

The behavior of the host in the state HH0 is defined as follows:

Take the following actions:
1. Generate a new ephemeral X25519 key pair (*host_ephemeral_privkey*, *host_ephemeral_pubkey*).
2. Send the message HandshakeInitiationRequest(*host_ephemeral_pubkey*) to the Trezor.
3. Transition to the state HH1(*host_ephemeral_privkey*, *host_ephemeral_pubkey*).

### State HH1

> The behavior of the host in the state HH1(*host_ephemeral_trezor_privkey*, *host_ephemeral_pubkey*) is defined as follows:

Reviewer 2: typo
```suggestion
The behavior of the host in the state HH1(*host_ephemeral_privkey*, *host_ephemeral_pubkey*) is defined as follows:
```

When the message HandshakeInitiationResponse(*trezor_ephemeral_pubkey*, *encrypted_trezor_static_pubkey*, *tag*) is received, take the following actions:
1. Set *h* = SHA-256(*protocol_name* || *device_properties*).
2. Set *h* = SHA-256(*h* || *host_ephemeral_pubkey*).
3. Set *h* = SHA-256(*h* || *try_to_unlock*).
4. Set *h* = SHA-256(*h* || *trezor_ephemeral_pubkey*).
5. Set *ck*, *k* = HKDF(*protocol_name*, X25519(*host_ephemeral_privkey*, *trezor_ephemeral_pubkey*)).

> 6. Set *trezor_masked_static_pubkey*, *success* = AES-GCM-DECRYPT(*key*=*k*, *IV*=0^96 (96 bits), *ad*=*h*, *plaintext*=*encrypted_trezor_static_pubkey*). Assert that *success* is True.

Reviewer: I would guess the connection / pairing is aborted or the process starts over, right?

When the message HandshakeCompletionResponse(*encrypted_trezor_state*) is received, take the following actions:
1. Set *key_request*, *key_response* = HKDF(*ck*, *empty_string*).
2. Set *trezor_state*, *success* = AES-GCM-DECRYPT(*key*=*key_response*, *IV*=*0^96*, *ad*=*empty_string*, *plaintext*=*trezor_state*). Assert that *success* is True.
3. Assert that *trezor_state* == STATE_UNPAIRED.
4. Set *nonce_request* = 0 and *nonce_response* = 1.
5. Set *encryption_state* = (*key_request*, *key_response*, *nonce_request*, *nonce_response*).
6. Transition to the state HP0(*pairing_methods*).

## Notes

The purpose of *prefix and padding* is to be compatible with the suite [CPACE-X25519-SHA512](https://www.ietf.org/archive/id/draft-irtf-cfrg-cpace-10.html#CPaceMontgomery) in the [symmetric setting](https://www.ietf.org/archive/id/draft-irtf-cfrg-cpace-10.html#name-responsibilities-of-the-app) as it was in [draft-irtf-cfrg-cpace-10](https://www.ietf.org/archive/id/draft-irtf-cfrg-cpace-10.html) where

- the associated data fields (ADa and ADb) are the empty strings,
- the session identifier (sid) and the channel identifier (CI) are the empty strings

> - the password-related string (PRS) is the code that the user rewrites from Trezor to the host and

Reviewer: I am not sure if it has been established already that user is to rewrite a code at some point. I know that
1. it was referenced at the beginning in protobuf definition of Pairing phase
2. we will learn about it literally in the next 3 paragraphs
3. it does not really matter
So my comment is probably nitpicking about clarity

Author: moved the Notes section to "after" the use of prefix and padding: https://github.com/trezor/trezor-firmware/commit/27fb661d90da61f130fe1f66178bc9165cb3fb73

    1. Generate a random 16-byte *code_entry_secret*.
    2. Set *commitment* = SHA-256(*code_entry_secret*).
    3. Send the message CodeEntryCommitment(*commitment*) to the host.
    4. Transition to the state TP2.
- selected_pairing_method is PairingMethod_CodeEntry and it WAS selected before
    1. Display *code_code_entry* on the Trezor’s screen (Note that *code_code_entry* is created only once per pairing - see State TP2)
    2. Send the message PairingPreparationsFinished() to the host.
    3. Transition to the state TP3a
- selected_pairing_method is PairingMethod_QrCode
    1. Generate a random 16-byte *qr_code_secret*.
    2.  Set *code_qr_code* = SHA-256(PairingMethod_QrCode || *handshake_hash* || *qr_code_secret*)[:16].
    3. Display QR Code on the Trezor’s screen
    4. Send the message PairingPreparationsFinished() to the host.
    5. Transition to the state TP3b
- selected_pairing_method is PairingMethod_Nfc
    1. Generate a random 16-byte *nfc_secret*.
    2. Display “NFC Dialog” on the Trezor’s screen
    3. Enable NFC (read-mode)
    4. Send the message PairingPreparationsFinished() to the host.
    5. Transition to the state TP3c

### State TP2

> The behavior of Trezor in the state TP2is defined as follows:

Reviewer: ```suggestion
The behavior of Trezor in the state TP2 is defined as follows:
```

- When the message CodeEntryChallenge(*challenge*) is received, take the following actions:
    1. Set *code_code_entry* = SHA-256(PairingMethod_CodeEntry || *handshake_hash* || *code_entry_secret* || *challenge* || *PairingMethod_CodeEntry*) % 1000000.
    2. Display *code_code_entry* on the screen.

>     3. Compute *pregenerator* as the first 32 bytes of SHA-512(*prefix* || *code_code_entry* || *padding* || *handshake_hash* || **0x00***), where *prefix* is the byte-string  0x08 || 0x43 || 0x50 || 0x61 || 0x63 || 0x65 || 0x32 || 0x35 || 0x35 || 0x06, *code_code_entry* is encoded as an ASCII representation of the 6-digit code (with prepended zeroes if needed) (`f"{code_code_entry:06}".encode("ascii")`), and *padding* is the byte-string **0x6f** || 0x00 ^ **111** || 0x20.

Reviewer: Grammatically-wise I think it is not pleasant to have two parenthesis sections right after each other. I would either drop the first one (as the second one describes the same) or combine them.

>     3. Compute *pregenerator* as the first 32 bytes of SHA-512(*prefix* || *code_code_entry* || *padding* || *handshake_hash* || **0x00***), where *prefix* is the byte-string  0x08 || 0x43 || 0x50 || 0x61 || 0x63 || 0x65 || 0x32 || 0x35 || 0x35 || 0x06, *code_code_entry* is encoded as an ASCII representation of the 6-digit code (with prepended zeroes if needed) (`f"{code_code_entry:06}".encode("ascii")`), and *padding* is the byte-string **0x6f** || 0x00 ^ **111** || 0x20.

Reviewer: Why are some of the values bold and others are not? I am especially confused by the last string `**0x6f** || 0x00 ^ **111** || 0x20`

- When the acknowledgment message is received and its sequence number equals 1, take the following actions:
    - Consider the previously sent message to be delivered.
    - Transition to the state S0b.

The behavior of the sender in the state S0b is defined as follows:

- If there is a message that needs to be sent, take the following actions:
    - Send the message with the sequence number equal to 1.
    - Switch to state S1a.
    - Reset and start the stopwatch.
- When the acknowledgment message is received and its sequence number equals 1, take the following actions:
    - Consider the message to be a duplicate acknowledgment of the previously sent message.
    - Transition to the state S0b.
- When the acknowledgment message is received and its sequence number equals 0, take the following actions:
    - Consider the message to be an acknowledgment received out of order.
    - Transition to the state S0b.

The behavior of the sender in the state S1 is analogous. The full finite-state transducer of the sender is depicted in the following diagram:

```mermaid
flowchart LR
    S0a["S0a"]
    S0b["S0b"]
    S1a["S1a"]
    S1b["S1b"]

    S0a -- receive Ack(Seq=0) --> S0a

>     S0b -- send Message(Seq=0, Payload=payload0) --> S1a

Reviewer: This `Seq` value seems inconsistent with the definition above. There you say
> ... S0b is defined as ...
> Send the message with the sequence number equal to 1.

But here from S0b you send sequence number 0.

Same for `S1b -- send Message(Seq=1, Payload=payload1) --> S0a`  I believe

    S1a -- receive Ack(Seq=1) --> S1a
    S1b -- send Message(Seq=1, Payload=payload1) --> S0a
    S1a -- receive Ack(Seq=0) --> S1b
    S1a -- after timeout\nresend Message(Seq=0, Payload=payload0) --> S1a:::dashed
    S1b -- receive Ack(Seq=*) --> S1b

    classDef dashed stroke-dasharray: 5 5;
```

The receiver alternates between states R0 and R1. The receiver starts in the state R0.

The behavior of the receiver in the state R0 is defined as follows:

- When the message is received and its sequence number equals 0, take the following actions:
    - Submits the received message to further processing.
    - Reply with an acknowledgement message with the sequence number equal to 0.
    - Transition to the state R1.
- When the message is received and its sequence number equals 1, take the following actions:
    - Consider the message to be a duplicate of the previously received message.
    - Reply with an acknowledgement message with the sequence number equal to 1.
    - Transition to the state R1.

The behavior of the receiver in the state R1 is analogous. The full finite-state transducer of the sender is depicted in the following diagram:

```mermaid
flowchart LR
    R1["R1"]

>     R2["R2"]

Reviewer: I believe this should be R0

>     R1 -- in_synchronization: Message(Seq=1, Payload=payload); out_synchronization: Ack(Seq=1) --> R1

Reviewer: what does `out_synchronization` and `in_synchronization` mean?

### State HH1

The behavior of the host in the state HH1(*host_ephemeral_privkey*, *host_ephemeral_pubkey*) is defined as follows:

When the message HandshakeInitiationResponse(*trezor_ephemeral_pubkey*, *encrypted_trezor_static_pubkey*, *tag*) is received, take the following actions:
1. Set *h* = SHA-256(*protocol_name* || *device_properties*).
2. Set *h* = SHA-256(*h* || *host_ephemeral_pubkey*).
3. Set *h* = SHA-256(*h* || *try_to_unlock*).
4. Set *h* = SHA-256(*h* || *trezor_ephemeral_pubkey*).
5. Set *ck*, *k* = HKDF(*protocol_name*, X25519(*host_ephemeral_privkey*, *trezor_ephemeral_pubkey*)).
6. Set *trezor_masked_static_pubkey*, *success* = AES-GCM-DECRYPT(*key*=*k*, *IV*=0^96 (96 bits), *ad*=*h*, *plaintext*=*encrypted_trezor_static_pubkey*). Assert that *success* is True.
7. Set *h* = SHA-256(*h* || *encrypted_trezor_static_pubkey*).
8. Set *ck*, *k* = HKDF(*ck*, X25519(*host_ephemeral_privkey*, *trezor_masked_static_pubkey*)).
9. Set *tag_of_empty_string*, *success* = AES-GCM-DECRYPT(*key*=*k*, *IV*=0^96, *ad*=*h*, *plaintext*=*empty_string*). Assert that *success* is True.
10. Set *h* = SHA-256(*h* || *tag_of_empty_string*).

> 11. Search *credentials* for a pairs (*trezor_static_pubkey*, *credential*) such that *trezor_masked_static_pubkey* == X25519(SHA-256(*trezor_static_pubkey* || *trezor_ephemeral_pubkey*), *trezor_static_pubkey*).

Reviewer: ```suggestion
11. Search *credentials* for pairs (*trezor_static_pubkey*, *credential*) such that *trezor_masked_static_pubkey* == X25519(SHA-256(*trezor_static_pubkey* || *trezor_ephemeral_pubkey*), *trezor_static_pubkey*).
```

Author: https://github.com/trezor/trezor-firmware/commit/e1c4d1775aa9ff1e31a9fea517cce033f1b10293

6. Transition to the state HP0(*pairing_methods*).

## Notes

The purpose of *prefix and padding* is to be compatible with the suite [CPACE-X25519-SHA512](https://www.ietf.org/archive/id/draft-irtf-cfrg-cpace-10.html#CPaceMontgomery) in the [symmetric setting](https://www.ietf.org/archive/id/draft-irtf-cfrg-cpace-10.html#name-responsibilities-of-the-app) as it was in [draft-irtf-cfrg-cpace-10](https://www.ietf.org/archive/id/draft-irtf-cfrg-cpace-10.html) where

- the associated data fields (ADa and ADb) are the empty strings,
- the session identifier (sid) and the channel identifier (CI) are the empty strings
- the password-related string (PRS) is the code that the user rewrites from Trezor to the host and
- *pairing_secret* is the resulting intermediate session key (ISK).

## Pairing phase

Pairing phase protects against potential MITM attacks. Currently, three pairing methods are designed:
- Code Entry - the user must re-write a code displayed on the Trezor to the host device.
- NFC - the host and the Trezor have a short communication using NFC, the host must be in proximity of the Trezor.
- QrCode - the user must scan a QR-Code from the Trezor's display.

> As of 2025/10/20, only CodeEntry pairing method is fully supported by Trezor devices. (See GitHub issue: [Add support for more THP pairing methods](https://github.com/trezor/trezor-firmware/issues/6036))

Reviewer: Is this still true?

  TP3b -- "in:&nbsp;QrCodeTag<br>out:&nbsp;QrCodeSecret" --> TC1
  TP3c -- "in:&nbsp;NfcTagHost<br>out:&nbsp;NfcTagTrezor" --> TC1
```

### State TP0

The behavior of Trezor in the state TP0 is defined as follows:

When the message PairingRequest(*host_name*, *app_name*) is received, take the following action:
1. Display a dialog "Allow {*app_name*} on {*host_name*} to pair with this Trezor?" on the screen.
2. Send ButtonRequest message.
3. Wait for ButtonAck Response from the host.
4. Wait for user interaction:
    1. The user cancelled the dialog - end pairing, send ActionCancelled Failure, drop the channel
    2. The user confirmed the dialog - transition to state TP1

### State TP1

The behavior of Trezor in the state TP1is defined as follows:

- When the message SelectMethod(selected_pairing_method) is received, transition to intermediate state “selected method”.

### State “selected method”

1. Read *selected_pairing_method* from the received message SelectMethod(*selected_pairing_method*)
2. Check that selected_pairing_method is allowed and supported by the Trezor. If it is not allowed, raise Exception

> - selected_pairing_method is PairingMethod_CodeEntry and it WAS NOT selected before

Reviewer: `selected_pairing_method` was written in cursive on line 869 but not in the list below

```suggestion
- *selected_pairing_method* is PairingMethod_CodeEntry and it WAS NOT selected before
```

Author: https://github.com/trezor/trezor-firmware/commit/08fa8ad54a29a7a79e3a9570061026ec1a5992fa
also added some of missing punctuation in the commit...

    5. Transition to the state TC1.

### Host’s state machine

```mermaid
flowchart TD
  HP0 -- "out: PairingRequest" --> HP1
  HP1 -- in: PairingRequestAck<br>out:&nbsp;SelectMethod[CodeEntry] --> HP2
  HP1 -- in: PairingRequestAck<br>out:&nbsp;SelectMethod[Nfc&nbsp;or&nbsp;QrCode] --> HP3b
  HP2 -- in: CodeEntryCommitement<br>out: CodeEntryChallenge --> HP3a
  HP3a -- in: CodeEntryCpaceTrezor --> HP4
  HP3b -- in: PairingPreparationsFinished --> HP4
  HP4 -- in:&nbsp;[user&nbsp;wants&nbsp;to&nbsp;change&nbsp;pairing&nbsp;method]<br>[selected method is CodeEntry and it was selected for the first time]<br> out: Select Method[CodeEntry] --> HP2
  HP4 -- in:&nbsp;[user&nbsp;wants&nbsp;to&nbsp;change&nbsp;pairing&nbsp;method]<br>[Nfc&nbsp;or&nbsp;QrCode&nbsp;or&nbsp;CodeEntry&nbsp;(not&nbsp;first&nbsp;time)]<br> out: Select Method --> HP3b
  HP4 -- in: [user enters code]<br>out:&nbsp;CodeEntryCpaceHostTag--> HP5
  HP5 -- in: CodeEntrySecret --> HC0
  HP4 -- in: [user scans qr code]<br>out: QrCodeTag --> HP6
  HP6 -- in: QrCodeSecret --> HC0
  HP4 -- in: [user uses NFC]<br>out: NfcTagHost --> HP7
  HP7 -- in: NfcTagTrezor --> HC0

> thpPairing

Author: Lonely it, indeed, was 😄 
Removed here: https://github.com/trezor/trezor-firmware/commit/8ff15408b0371959e89fa00a02eb88e53838244d

}

message PairingCredential {
    option (internal_only) = true;
    required CredentialMetadata cred_metadata = 1;    // Credential metadata
    required bytes mac = 2;                           // Message authentication code generated by the Trezor
}

message AuthenticatedCredentialData {
    option (internal_only) = true;
    required bytes host_static_public_key = 1;        // Host's static public key used in the handshake
    required ThpCredentialMetadata cred_metadata = 2; // Credential metadata
}
```

### Credential authentication key

For credential issuance and validation, a credential authentication key (*cred_auth_key*) is used. This key is derived from two components - *device secret* and a counter (*cred_auth_key_counter*). The *device secret* is independent of seed and is generated randomly first time it is needed. The *cred_auth_key_counter* is 4 bytes big-endian initiated to zero.

### Credential validation

Let *ValidateCredential*(*cred_auth_key*, *credential_binary*, *host_static_pubkey*) be the function defined as follows:

1. Set *credential* = PROTOBUF-DECODE(*type*=*PairingCredential*, *credential_binary*).

> 2. Set *authenticated_credential_data* = PROTOBUF-ENCODE(*type*=AuthenticatedCredentialData, *host_static_pubkey*, *credential.cred_metadata*).

Reviewer: Why do we `PROTOBUF-ENCODE` the date here?

3. Set *mac* = HMAC-SHA-256(*cred_auth_key*, *authenticated_credential_data*).
4. Return *mac* == *credential.mac*.

### Credential invalidation

Incrementing the *cred_auth_key_counter* results in a new *cred_auth_key*, invalidating all previously issued THP credentials. The design does not allow to invalidate a single selected credential - only all credentials can be invalidated, or none.

### Credential issuance

Let *IssueCredential*(*cred_auth_key*, *host_static_pubkey*, *credential_metadata*) be the function defined as follows:

1. Set *authenticated_credential_data* = PROTOBUF-ENCODE(*type*=AuthenticatedCredentialData, *host_static_pubkey*, *credential_metadata*).
2. Set *mac* = HMAC-SHA-256(*cred_auth_key*, *authenticated_credential_data*).
3. Set *credential_raw* = PROTOBUF-ENCODE(*type*=*PairingCredential*, *credential_metadata*, *mac*).
4. Return *credential_raw*.

### Autoconnect credential issuance

> The host can request an autoconnect credential by sending the *ThpCredentialRequest* message to the Trezor - it is even possible to do when the communication is in the encrypted transport phase. The request must contain an already issued (valid) credential. Trezor does not allow host to request an *autoconnect* credential without the host providing **a** valid credential.

Reviewer: Is the confirmation screen for autoconnect different from the confirmation screen for issuing the non-autoconnect credential?
